### PR TITLE
recharge.iced obtains the transaction fee estimate from an api

### PR DESCRIPTION
**Pull request by heronhaye**

### CODE REVIEW

File: src/recharge.iced [hunk 1]
```diff
@@ -80,8 +81,46 @@ exports.Runner = class Runner extends Base
...
+    else if opts.type == 'litecoin'
+      cb null, 100
+    else
+      cb new Error("Unknown cryptocurrency " + opts.type), 0
+
+  # Estimates fee needed to send a transaction based on
+  # the parameters in @marginal_fee_estimator, capped by
+  # opts.feePerByteLimit and multiplied by opts.padding.
+  # No default parameters set.
+  fee_estimator : (opts, cb) ->
```
Comment (maxtaco) [inline src/recharge.iced]: How about?

```iced
fee_estimator : ({type, feePerByteLimit, tx}, cb) ->
```

File: src/recharge.iced [hunk 2]
```diff
@@ -93,7 +132,16 @@ exports.Runner = class Runner extends Base
     change_offset = tx.addOutput @change_address, 1
     tx.sign 0, skey
 
-    fee = btcjs.networks.bitcoin.estimateFee(tx)
+    # Can change these settings...
+    btc_opts = {
```
Comment (maxtaco) [inline src/recharge.iced:98]: Should somehow read this out of the config file in the production (non-test) case.

**Review by maxtaco — CHANGES_REQUESTED** [review]
Just as a sanity check, how many times is the transaction fee estimate called for making an N-sized recharge?  I would expect (N+1), in which case it might make sense to only make ~2 API calls to 21.co.  I'm worried we'll run up against their rate limit otherwise.

**Comment by heronhaye** [comment]
Now the fee estimation is done when initializing base and the @marginal_fee is stored as a class variable for use by any subclass without calling the API multiple times per run of the script.

**Review by maxtaco — APPROVED** [review]

File: src/base.iced [hunk 3]
```diff
@@ -100,13 +111,64 @@ exports.Base = class Base
...
+  # Estimates fee per byte for a specific network type needed to achieve
+  # verification before maxClearanceMins minutes For bitcoin, returns in
+  # satoshis and uses the 21.co API with no fallback.
+  marginal_fee_estimator: ({type, maxClearanceMins}, cb) ->
+    esc = make_esc cb,'marginal_fee_estimator'
+    if type == 'bitcoin'
+      apiUrl = 'https://bitcoinfees.21.co/api/v1/fees/list'
+      await request apiUrl, esc defer resp, body
+      if resp.statusCode == 200
+        fees = JSON.parse(body)['fees']
```
Comment (maxtaco) [inline src/base.iced]: actually can you make this a little more robust?  `JSON.parse` can throw, and `fees` might be undefined. In either case the program will crash.

**Review by maxtaco — CHANGES_REQUESTED** [review]
One small change

**Comment by heronhaye** [comment]
Some different ways this could be handled,
-Use a schema library
-Move aget/aassert to a util library in this package
-Move aget/aassert to iced-utils

File: src/base.iced [hunk 4]
```diff
@@ -135,13 +145,17 @@ exports.Base = class Base
       apiUrl = 'https://bitcoinfees.21.co/api/v1/fees/list'
       await request apiUrl, esc defer resp, body
       if resp.statusCode == 200
-        fees = JSON.parse(body)['fees']
+        await a_json_parse body, esc defer body_json
+        await @aget body_json, 'fees', esc defer fees
```
Comment (maxtaco) [inline src/base.iced]: well i've never seen this done before, but it works.  i guess if i were doing this i would probably write:

```iced
currentClearanceMins = 10000
for {maxMinutes,MaxFee} in (body_json.fees or []) when maxMinutes <= currentClearanceMins
  currentClearanceMins = maxMinutes
  currentFee = maxFee
  
```

**Review by maxtaco — APPROVED** [review]

**Pull request merged**